dex/farm-with-locked-rewards/src/lib.rs

```diff
@@ -0,0 +1,331 @@
+        let farm_token_id = self.farm_token().get_token_id();
+        self.burn_farm_tokens_from_payments(&payments);
+
+        let new_tokens =
+            self.mint_farm_tokens(farm_token_id, attrs.current_farm_amount.clone(), &attrs);
+
+        self.send_payment_non_zero(&caller, &new_tokens);
+
+        new_tokens
+    }
+
+    #[endpoint(startProduceRewards)]
+    fn start_produce_rewards_endpoint(&self) {
+        self.require_caller_has_admin_permissions();
+        self.start_produce_rewards();
+    }
+
+    #[endpoint]
+    fn end_produce_rewards(&self) {
+        self.require_caller_has_admin_permissions();
+
+        let mut storage = StorageCache::new(self);
+
+        self.generate_aggregated_rewards_with_boosted_yields(&mut storage);
+        self.produce_rewards_enabled().set(false);
+    }
+
+    #[endpoint(setPerBlockRewardAmount)]
+    fn set_per_block_rewards(&self, per_block_amount: BigUint) {
+        self.require_caller_has_admin_permissions();
+        require!(per_block_amount != 0u64, ERROR_ZERO_AMOUNT);
+
+        let mut storage = StorageCache::new(self);
+
+        self.generate_aggregated_rewards_with_boosted_yields(&mut storage);
+        self.per_block_reward_amount().set(&per_block_amount);
+    }
+
+    fn generate_aggregated_rewards_with_boosted_yields(&self, storage_cache: &mut StorageCache<Self>) {
+        let mint_function = |token_id: &TokenIdentifier, amount: &BigUint| {
+            self.send().esdt_local_mint(token_id, 0, amount);
+        };
+        let total_reward =
+            self.mint_per_block_rewards(&storage_cache.reward_token_id, mint_function);
+
+        if total_reward > 0u64 {
+            storage_cache.reward_reserve += &total_reward;
+            let split_rewards = self.take_reward_slice(total_reward);
+
+            if storage_cache.farm_token_supply != 0u64 {
+                let increase = (&split_rewards.base_farm * &storage_cache.division_safety_constant)
+                    / &storage_cache.farm_token_supply;
+                storage_cache.reward_per_share += &increase;
+            }
+        }
+    }
+
+    fn calculate_reward_with_boosted_yields(
+        &self,
+        user: &ManagedAddress,
+        farm_token_amount: &BigUint,
+        attributes: &FarmTokenAttributes<Self::Api>,
+        storage_cache: &StorageCache<Self>)
+        -> BigUint
+    {
+        let base_farm_reward = self.default_calculate_reward(farm_token_amount, attributes, storage_cache);
+        let boosted_yield_rewards = self.claim_boosted_yields_rewards(user, &storage_cache.reward_token_id);
+        base_farm_reward + boosted_yield_rewards
+    }
+
+    fn require_queried(&self) {
+        let caller = self.blockchain().get_caller();
+        let sc_address = self.blockchain().get_sc_address();
+        require!(caller == sc_address, "May only call this function through VM query");
+    }
+
+    fn send_to_lock_contract_non_zero(
```

Comment:
This function should return the output payment received from `lock_tokens_and_forward`, since you'll need it to propagate it as endpoint result.